doc/high-frequency-telemetry/high-frequency-telemetry-hld.md

```diff
@@ -0,0 +1,706 @@
+160-160: "1"
+161-175: "Element ID = 5 (port index)"
+176-191: "Field Length = 8 bytes"
+192-223: "Enterprise Number = 0x00010004"
+
+```
+
+#### 7.2.3. IPFIX data
+
+An IPFIX data message consists of snapshots that is a binary block that can be interpreted using the IPFIX template mentioned above.
+
+The binary structure of a snapshot is as follows:
+
+``` mermaid
+
+---
+title: A snapshot of IPFIX data
+---
+packet-beta
+0-15: "Set ID = Same as template ID"
+16-31: "Set Length = (4 + 8 + Number of stats * 8) bytes"
+32-95: "Data 1: observationTimeNanoseconds"
+96-127: "Data 2: Stats 1"
+128-159: "..."
+160-191: "Data N + 1: Stats N"
+```
+
+- The snapshot structure is derived from the IPFIX template, which is based on the stats we want to record.
+
+Below is an example of an IPFIX message with 3 snapshots for the same stats record as the IPFIX template example:
+
+``` mermaid
+
+---
+title: stream message IPFIX
+---
+packet-beta
+0-15: "Version = 0x000a"
+16-31: "Message Length = 112 bytes"
+32-63: "Export Timestamp = 2024-08-29 20:30:60"
+64-95: "Sequence Number = 1"
+96-127: "Observation Domain ID = 0"
+
+128-143: "Set ID = 256"
+144-159: "Set Length = 36 bytes"
+160-223: "observationTimeNanoseconds = 10000"
+224-287: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 10"
+288-351: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+352-415: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 5"
+
+416-431: "Set ID = 256"
+432-447: "Set Length = 36 bytes"
+448-511: "observationTimeNanoseconds = 20000"
+512-575: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 15"
+576-639: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+640-703: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 6"
+
+704-719: "Set ID = 256"
+720-735: "Set Length = 36 bytes"
+736-799: "observationTimeNanoseconds = 30000"
+800-863: "Port 1: SAI_PORT_STAT_IF_IN_ERRORS = 20"
+864-927: "Port 2: SAI_PORT_STAT_IF_IN_ERRORS = 0"
+928-991: "Port 3: SAI_PORT_STAT_IF_IN_ERRORS = 8"
+
+```
+
+- If the number of stats in a group is small, multiple snapshots may be encoded into a single IPFIX message.
+- If the number of stats in a group exceeds 8K, the group must be split across multiple IPFIX messages.
+
+The IPFIX template should be provided by vendors. This document does not restrict how to split or concatenate snapshots, but each separated snapshot must include its own `observationTimeNanoseconds`.
+
+#### 7.2.4. Netlink message
+
+We expect all control messages and out-of-band information to be transmitted by the SAI. Therefore, it is unnecessary to read the attribute header of netlink and the message header of Genetlink from the socket. Instead, we can insert a bulk of IPFIX recordings as the payload of the netlink message. The sample code for building the message from the kernel side is as follows:
+
+``` c
+
+struct genl_multicast_group stel_mcgrps[] = {
+    { .name = "ipfix" },
+};
+
+// Family definition
+static struct genl_family stel_family = {
+    .name = "sonic_stel",
+    .version = 1,
+    // ...
```

Comment:
Please specify if any genetlink attribute is to be defined for the IPFix message. Providing a attribute will help us to extend the genetlink to provide additional info other than IPFix messages.

Reply:
There is no any genetlink attribute in this interface. This design is intentional, because we want all extensions to go through SAI and not other channels.